### ISSUE TITLE
insert queries having auto-generated ids are not executed

### ISSUE THREAD

**Issue by oriy** [jOOQ/jOOQ#6935]
### Expected behavior and actual behavior:

_expected_: insert query to be executed, new id to be returned

_actual_: insert query is not executed, as identified as **not executable**

### Steps to reproduce the problem:

table with auto-generated fields only
```
CREATE TABLE `my_table` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `created_timestamp` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB
```

Using jOOQ generated DAO and POJO

```
myTableDao.count() // 0
myTableDao.insert(new MyTable())
myTableDao.count() // 0
```

InsertQuery object:
```
insert into `my_table`\ndefault values
```

expected query to be executed:
```
insert into my_table() values ();
```

### Versions:

- jOOQ:
jOOQ version:3.10.1
- Database (include vendor):
MySQL

----
See also: https://github.com/jOOQ/jOOQ/issues/10282

**Comment by lukaseder** [comment]
Thank you very much for your report. I will look into this shortly.

**Comment by lukaseder** [comment]
I cannot reproduce this out of the box with jOOQ 3.11.0-SNAPSHOT and 3.10.2. The query that gets generated is:

```sql
insert into `test`.`t_6935` (`created_timestamp`) values (null)
```

Do you have any additional info you could share, e.g. about your code generation configuration?

In any case, it's very interesting to see that MySQL supports this non-standard way to express the `default values` clause. This is definitely worth implementing as an emulation of `default values`. I have created a separate issue for this: #6936

**Comment by oriy** [comment]
Thanks @lukaseder for the extremely fast response

My sample was misleading,
Original table has id only (which I'm not prowd of...)

Please recheck this usecase

**Comment by lukaseder** [comment]
Ah yes, then I can reproduce this and it is in fact the same thing as #6936. Should be an easy fix

**Comment by lukaseder** [comment]
No it's not the same issue as #6936. For some reason, the internal `InsertQueryImpl` is marked as "not executable", for which `DEFAULT VALUES` is the default generated SQL in the logs, but we're not actually executing anything:

![image](https://user-images.githubusercontent.com/734593/34157461-e20eda2c-e4c1-11e7-822e-b621e818cb24.png)

Will further investigate this

**Comment by lukaseder** [comment]
I see. An `UpdatableRecord` without a single changed value is currently not stored when `insert()` is called upon it. This might make sense for records that have been previously fetched from the database, and then `store()` is called on them. But in this case, the expectation is clearly to have the record inserted with default values only. This is specifically true as adding the timestamp column changes the behaviour.

I'll need to think about the correct fix a bit more thoroughly. This might turn into a breaking change in `UpdatableRecord` behaviour...

#6936 is an easy fix but not strictly related...

**Comment by oriy** [comment]
on scala, my teammate's workaround was replacing: 

```
db
 .insertInto(MY_TABLE)
 .returning(MY_TABLE.ID)
 .fetchOne()
 .get(MY_TABLE.ID)
```
with 
```
db
 .insertInto(MY_TABLE, MY_TABLE.ID)
 .values(null.asInstanceOf[Integer])
 .returning(MY_TABLE.ID)
 .fetchOne()
 .get(MY_TABLE.ID)
```

**Comment by lukaseder** [comment]
Yes, that works as now you're explicitly setting a `NULL` value (which will be overridden by the default in MySQL) rather than implicitly using the `DEFAULT` value.

**Comment by lukaseder** [comment]
Same issue: https://github.com/jOOQ/jOOQ/issues/10282

**Comment by lukaseder** [comment]
Looking into this now. For the record, the ideal way to implement this with a query would be to use the `DEFAULT VALUES` clause:

```java
db
 .insertInto(MY_TABLE)
 .defaultValues()
 .returning(MY_TABLE.ID)
 .fetchOne()
 .get(MY_TABLE.ID)
```

But this should also work with DAOs, where the query is not accessible.

**Comment by lukaseder** [comment]
I've given this and #10282 some additional thought. While #10282 happens with all tables that have an identity, this one here is an edge case where:

- A table has only one column
- That column is a primary key *and* an identity
- You're using a DAO

#10282 would allow for some additional convenience to be able to call `TableRecord.insert()` and automatically fetch the generated ID again, in a single line of code, this would automatically translate to your DAO usage where the only difference is that you don't have any extra columns.

So, with DAOs, it's a bit different because they operate on POJOs, not records. POJOs don't have `changed` flag, so we always assume everything is `changed`.

Hence, we should have an internal flag that overrides the "executable" check in `TableRecord`. If we decide to go forward with #10282, this internal flag could be made publicly available.

**Comment by lukaseder** [comment]
However, if this is implemented first (`DAO` behaviour is changed) and #10282 is implemented later (`Record` behaviour is changed), then there are two possibly backwards incompatible changes in a row.

Better wait with fixing anything until a decision is made for both cases.

**Comment by lukaseder** [comment]
This issue will be fixed implicitly via #10386

**Comment by lukaseder** [comment]
This is now fixed via #10386, which changes the default for `TableRecord.insert()`. Remaining topics will be dealt with in #10386.
